data_gateway/serial_protocol.py

```diff
@@ -0,0 +1,146 @@
+
+    def data_received(self, data):
+        """Pull all waiting data from the serial port into an in-memory buffer, assemble it into packets, and handle
+        the packets.
+
+        :param bytes data:
+        :return None:
+        """
+        if len(data) == 0:
+            return
+
+        self.buffer.extend(data)
+
+        while len(self.buffer) > 0:
+
+            byte = bytearray([self.buffer.pop(0)])
+
+            # Start collecting a packet if the initiator byte is received.
+            if byte == self.initiator:
+                self._currently_receiving_packet = True
+                continue
+
+            # Ignore byte if not currently receiving a packet.
+            if not self._currently_receiving_packet:
+                continue
+
+            # Receive further bytes for the current packet.
+            if self.current_packet_type is None:
+                self.current_packet_type = str(int.from_bytes(byte, self.transport.packet_reader.config.endian))
+                continue
+
+            # Get the expected length of the packet.
+            if self.current_packet_expected_length is None:
+                self.current_packet_expected_length = int.from_bytes(byte, self.transport.packet_reader.config.endian)
+                continue
+
+            # Collect more data until the expected packet length is reached.
+            if len(self.current_packet) < self.current_packet_expected_length:
+                self.current_packet.extend(byte)
+
+            # When the expected packet length is reached, handle the packet and get ready for the next one.
+            if len(self.current_packet) == self.current_packet_expected_length:
+                self.handle_packet()
+                self._prepare_for_new_packet()
+
+            if self.transport.stop_when_no_more_data and self.transport.serial.in_waiting == 0:
+                self.transport.serial.close()
+                self.transport.packet_reader.stop = True
+
+    def handle_packet(self):
+        """Handle a packet by either updating the handles or parsing it as a payload for the packet reader.
+
+        :return None:
+        """
+        if self.current_packet_type == str(self.transport.packet_reader.config.type_handle_def):
+            self.transport.packet_reader.update_handles(self.current_packet)
+            return
+
+        self.transport.packet_reader.parse_payload(
+            packet_type=self.current_packet_type,
+            payload=self.current_packet,
+            data=self.collected_data,
+            previous_timestamp=self.previous_timestamp,
+        )
+
+    def _prepare_for_new_packet(self):
+        """Reset the current packet information.
+
+        :return None:
+        """
+        self.current_packet = bytearray()
+        self.current_packet_type = None
+        self.current_packet_expected_length = None
+        self._currently_receiving_packet = False
+
+
+class CustomReaderThread(serial.threaded.ReaderThread):
```

Comment:
I’m not a big fan of calling subclasses ‘Custom<Thing>’ because that’s self evident from the fact that you subclass it. Is there a more meaningful name that tells us what the subclass does specially?